### ISSUE TITLE
HMM unable to download from GitHub

### ISSUE THREAD

**Issue by Dupiter-AU** [sonic2kk/steamtinkerlaunch#677]
## System Information
- SteamTinkerLaunch version: `v11.12.20221124-1`
- Distribution: `Arch Linux`
- Installation Method: `AUR: yay`

## Issue Description

While attempting to install HMM, STL is unable to connect to GitHub; `WARNING: Can't reach GitHub - Not attempting to download HedgeModManager`
I have attempted reinstall, deleting the files in `~/.config/steamtinkerlaunch/`, and checking the logs didn't prove any use to me.
I have a working internet connection and can connect to GitHub just fine.

## Logs

[steamtinkerlaunch.log](https://github.com/sonic2kk/steamtinkerlaunch/files/10118526/steamtinkerlaunch.log)

I don't know what other logs could help, or where. After looking in `~/.config/steamtinkerlaunch/logs/`, I don't _think_ there are any relevant logs?

**Comment by sonic2kk** [comment]
Very odd, another user recently reported that SteamTinkerLaunch could not download some necessary packages from the Arch repos.

This was due to a DNS issue, maybe it's the same on your end? I can't reproduce the problem on my PC (also running Arch Linux) or Steam Deck, it can connect and download just fine.

As a temporary workaround you could manually download the HMM exe to `~/.config/steamtinkerlaunch/downloads/hedgemodmanager` but that's a bit hacky. Since it can't connect to GitHub it also won't be able to store the version, and I'm not sure what Unforeseen Consequences™ might arise from that.

Also, the log you provided was fine :-) Thank you

**Comment by Dupiter-AU** [comment]
Interesting that STL wouldn't download some dependencies from the repos. Do these dependencies look OK to you?

![Screenshot_20221130_120736](https://user-images.githubusercontent.com/82429716/204690103-48f297fa-2014-4f4b-b7fc-4512183ee2e4.png)

Assuming that not having the version causes "Unforeseen Consequences™", Would it be possible to manually install the version as well in that case?

I'm afraid I don't entirely know what DNS is, but I'm certain that my internet connection works perfectly, especially as other programs are able to successfully install from the internet.

**Comment by Dupiter-AU** [comment]
Hold on, the manual install has changed a lot of stuff, give me a moment and I'll send the new log

**Comment by Dupiter-AU** [comment]
[steamtinkerlaunch.log](https://github.com/sonic2kk/steamtinkerlaunch/files/10118774/steamtinkerlaunch.log)

Looks like the install was successful, but HMM never started so I'm guessing something's wrong on that part.

EDIT 1: `Wed 30 Nov 2022 12:15:48 AEST WARN - dlLatestHMM - No HMMVERFILE found at '/home/dupiter/.config/steamtinkerlaunch/downloads/hedgemodmanager/hmmver.txt' - HedgeModManager may not have been installed and so may fail to starrt offline!` looks like this might be the leading cause of starting failures, but I'm not sure

**Comment by sonic2kk** [comment]
> Interesting that STL wouldn't download some dependencies from the repos.

Ah, sorry, I meant on Steam Deck for that user (it's one of the recent comments in the Steam Deck support issue). It fetches some packages from the Arch archives, and for this user they couldn't correctly ping archlinux.org on their home(?)  connection, but it worked on their college connection.

The issues seem similar to me, but I could be totally off. The logic for warning about Steam Deck dependencies and for this HMM GitHub warning are identical - It's just that HMM checks if it can ping to GitHub before attempting to download instead of archlinux.org.

> Would it be possible to manually install the version as well in that case?

Yes, you can create a text file with the version string in the same folder as the HMM exe (`~/.config/steamtinkerlaunch/downloads/hedgemodmanager`). I think the file is supposed to be called `hmmvers.txt`, and you paste your version in like `7.2-8` or the commit sha for the nightly.

DNS is, ***very*** simply (because I don't understand the full ins and outs myself) how your computer looks up the IP to connect to a server. It's entirely possible that a browser can connect or even that some other programs work, but that others do not. Sometimes this is a temporary ISP issue, a temporary router issue, or in general just bad luck.

I don't think there's anything on the SteamTinkerLaunch side that is broken, nothing has been touched in almost a week (I'm currently taking a bit of a break 😅) and I can't replicate the error currently. It seems to me like it could be a network issue, but I won't rule out that this could be a fault with SteamTinkerLaunch just yet.

That warning is triggered when SteamTinkerLaunch attempts to ping GitHub but gets an error status code back. There isn't anything too fancy going on, so it seems that part of the code is either being incorrectly triggered or there's some kind of network issue.

It's late for me so I don't have time to dig in too much to this right now, but fingers crossed you get somewhere with a manual download!

**Comment by sonic2kk** [comment]
Just skimmed the log on my phone (so I may not have it right) but it seems like HMM isn't correctly catching the HedgeModManager executable.

That was my bad for the file, it's `hmmver.txt` - Sorry about that!

SteamTinkerLaunch should expect the exe to be called `HedgeModManager.exe`, I believe this is the default name. Just in case it can't find the exe with a different name.

**Comment by Dupiter-AU** [comment]
Hmm, my manual install filename _is_ `HedgeModManager.exe`, and the `hmmver.txt` file exists with the version name inside (7.8-2), but still nothing is happening. HMM still does not launch. 

Don't dig into this right now, if it's late where you are, you should probably be going to bed. Goodnight!

**Comment by Dupiter-AU** [comment]
One other thing I should mention: my PC is _**REALLY**_ weird when it comes to WINE applications.
For some reason, WINE applications on my PC can't ever find ANY hidden files (`/.` prefix). I have no idea why and I don't know how to fix it. My temporary solution with other applications was to make symlinks but I don't think I can do that here.
This may be the cause of the problem with launching HMM, but I have no idea how to solve that.

**Comment by sonic2kk** [comment]
Wine is a complicated beast! Maybe this is part of the issue indeed. I'll give some background on how the process works (before I do actually go to bed 😅)

SteamTinkerLaunch doesn't use your system Wine for HMM, it uses the latest GE-Proton in your Steam compatibility tools folder, or downloads one if it doesn't exist (it downloads it to somewhere in `~/.config/steamtinkerlaunch/`, I forget the exact path).

It runs the Wine binary for that Proton build to set up a 64bit prefix for HMM to run in at `~/.config/steamtinkerlaunch/hedgemodmanager/compatdata/pfx`. The reason for this is that HMM needs a few Winetricks to run: `dotnet48`, `vcrun2019`, `d3dcompiler_47` and `d3dx9` (not too sure on the name of the last one -- it's something to that effect). The dedicated HMM prefix is created with a GE-Proton Wine and the Winetricks are installed with this build, as regular Proton and Proton Experimental don't seem to be able to install `dotnet48`.

Something all of these paths have in common is that they're all in hidden folders! So if you've run into issues with these before, you may need to create symlinks here. Sorry for this inconvenience but I hope those paths help you with trying to make symlinks -- If this is indeed the problem! 

One thing I would check first is if the HMM prefix is being correctly created at all. You could check the folder size and/or have a browse to see if the prefix structure looks at least somewhat correct. Though from the log that I skimmed it seemed like it wasn't even trying to create the prefix for some reason.

**Comment by Dupiter-AU** [comment]
Nope! STL is not creating a 64bit prefix at `~/.config/steamtinkerlaunch/hedgemodmanager/compatdata/pfx`. Not even the folder itself.
I have, however just found a checkbox in winecfg to show hidden files. This might be a good fix, but I would need to run winecfg in the same prefix that STL is running in. Which I don't know.

**Comment by Dupiter-AU** [comment]
Yeah... I've got nothing.

Tried opening a winecfg window into the default Proton-GE prefix, but couldn't
Tried adding "Show hidden directories" to my default winecfg and ran a program through that. Still couldn't access hidden files.
I doubt making a symlink would work here, as all the directories that STL creates are automatic and already set, so it wouldn't re-route anyway.

I don't know if this is a WINE problem or if my PC is causing some weird problem down the line.

If worse comes to worse, would it be possible to _manually_ install HMM?

**Comment by sonic2kk** [comment]
> Not even the folder itself.

This would be why HMM is not opening then :-) There is no prefix for it to run in! I guess this is something that we'll need to investigate.

<hr>

I went and checked the code -- The offline check appears to be correct. The code STL runs for the check is 

```bash
# Internet connection check
if ! ping -q -c1 github.com &>/dev/null; then
	writelog "WARN" "${FUNCNAME[0]} - Looks like we're offline or GitHub is down, not attempting to download HedgeModManager when offline - May cause issues if no HMM exe is downloaded!"
	writelog "WARN" "${FUNCNAME[0]} - Will still attempt to install Winetricks as they may be cached"
	echo "WARNING: Can't reach GitHub - Not attempting to download HedgeModManager"

	# Set HMMVER offline if we have it in the HMMVERFILE + some warning logging if there is no version file
	if [ -f "$HMMVERFILE" ]; then
		HMMVER="$( head -n 1 "$HMMVERFILE" )"
		if [ -z "$HMMVER" ]; then
			writelog "WARN" "${FUNCNAME[0]} - HedgeModManager version stored in HMMVERFILE at '$HMMVERFILE' appears to be empty, installation may have failed last time!"
		else
			writelog "INFO" "${FUNCNAME[0]} - HedgeModManager version stored in HMMVERFILE is '$HMMVER' - It seems like HedgeModManager successfully installed before, so running offline should work fine"
		fi
	else
		writelog "WARN" "${FUNCNAME[0]} - No HMMVERFILE found at '$HMMVERFILE' - HedgeModManager may not have been installed and so may fail to starrt offline!"
	fi

	return  # The code below this is for downloading HMM, which we don't attempt when we can't reach GitHub, so we return
fi
```

However, inside the offline code block, I noticed an error when STL tries to get the HMM version. I was checking `[ -d "$HMMVERFILE" ]`, which is a flag for checking if a directory exists. This should have been `-f`, and was fixed with commit 7a02fda8aaec118faab20f01013f34593e5cde9b.

This error doesn't explain why the offline check is failing, this is just a single check that is done in the offline block. As a test, from a Bash prompt, you could run `if ! ping -q -c1 github.com &>/dev/null; then echo "Bash can't reach GitHub"; else echo "Bash can reach GitHub"; fi` -- This will print out to the console a relevant error message. This is exactly the same code that SteamTinkerLaunch runs, so the result for STL and your command prompt should be the same. Because the core issue here is that HMM can't be downloaded :-)

Another issue that I found is in `installHMM` - the function that manages setting up the prefix. It falls over automatically if there is no file at the path where the HMM exe should be stored, and it appears to me that this value (stored in variable `HMMSPATH`) is _only_ stored when Hedge Mod Manager is downloaded. There is a line in your log like `SKIP - installHMM - '' not found - Nothing to install - skipping` - The `''` here is the `HMMSPATH` variable, and this was blank! I attempted to fix this issue with 211292f452681ae32381b19cd6a163f0441f88d6.

So the offline issue shouldn't be happening for you to begin with, but also regardless your tests here helped me find some pretty critical bugs with the offline process, so thank you :-) And thanks in general for taking the initiative to try other steps too. I am hoping now that HMM should correctly detect your executable and set itself up correctly to create the prefix.

You sound like you have some experience with Wine already but just as a courtesy heads up: If you've never installed `dotnet48` into a prefix before, it can take *a very* long time. It really depends, usually for me it takes around 5 minutes, but at worst it can take up to 10 minutes! This is, as far as I am aware, outside of STL's control and is just a Winetricks thing. On top of `dotnet48`, STL installs some other Winetricks, which should be faster but may still take some time (i.e., `vcrun2019` may take some time). All this to say, it make take some time for HMM to start up, since now it will actually be """installed""" (Actually, HMM doesn't "install" at all, it's a standalone exe, but STL calls the the prefix setup the "Install" process 😄)

<hr>

The tl;dr here is that I've pushed two commits, and if you are using SteamTinkerLaunch v11.12.20221130-2, you should be able to install Hedge Mod Manager offline -- Or at least SteamTinkerLaunch should get further.

Thanks for the patience, and fingers crossed!

**Comment by Dupiter-AU** [comment]
Well, it seems like my bash cannot reach github, as I ran `if ! ping -q -c1 github.com &>/dev/null; then echo "Bash can't reach GitHub"; else echo "Bash can reach GitHub"; fi` and got the same error message. I don't know what's causing that...
![image](https://user-images.githubusercontent.com/82429716/204938944-3b44c5cc-e613-4741-8891-c88ab08f8024.png)

Yes, I do have a bit of experience with wine. I have a program called Hammer and Hammer-Plus-Plus used for making Maps for Source engine games, and that took a little bit of tinkering to set up. Funny enough, I have used winetricks to install dotnet48 and multiple vcruns for that exact thing (with some errors of course)! This is also the program I used to test being able to access hidden files from a wine application, however I could not access hidden files.

I would love to use the updated STL version, but it does not appear to be on the AUR yet, so I guess I'll have to wait a bit for that.

Thanks for the help!

**Comment by sonic2kk** [comment]
That is very interesting about the same error happening from the command line. I guess this is some network issue external to SteamTinkerLaunch -- not to say that this is *your* user error, this could very well be a case of an ISP being dumb :sweat_smile: This was the case with another issue

I'm not sure if anything can be done about this from the SteamTinkerLaunch end. Though it would be interesting if you could try running this command with other domains (e.g. `archlinux.org`, `google.com`) and as well as this, just simply running `ping github.com` (or as well, any other domains you want to try :smiley:)

Another very interesting test would be if you could test that Bash line with another internet connection and/or on another device -- For example, mobile data, another location, or even with a VPN on (or off, if you're using one). But of course, there is no pressure to do this. To be honest it's mostly curiosity on my part.

If the ping fails, other commands that SteamTinkerLaunch executes later will almost certainly fail as well, which is why the check is in place. Perhaps I could consider adding a `--force` flag or something, to keep going even in the case of an offline check failing... Not sure yet :-) I will think more on how to go forward and probably wait to see if this affects more users.

>  Funny enough, I have used winetricks to install dotnet48 and multiple vcruns for that exact thing (with some errors of course)!

Hehe, yeah the whole Winetricks thing (*particularly* dotnet48) has been a pain with HMM as for some users dotnet48 randomly does not work (there is an open issue where dotnet40 and dotnet48 are failing to install with HMM on SteamOS, seems to be a sporadic winetricks issue that I still cannot figure out >_<;)

> I would love to use the updated STL version, but it does not appear to be on the AUR yet, so I guess I'll have to wait a bit for that.

Ah, sorry about that. I'm not heavily involved in the AUR package (I actually only took over as project lead on SteamTinkerLaunch a little over a month ago), but if you would like you could install SteamTinkerLaunch from source. You can download  the repo and run `sudo make install` - which is basically the same thing the AUR package does if you compare [its PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=steamtinkerlaunch-git#n49) to the [SteamTinkerLaunch Makefile](https://github.com/sonic2kk/steamtinkerlaunch/blob/master/Makefile#L9-L18). The Makefile also has an `uninstall` block, so you can remove SteamTinkerLaunch by running `sudo make uninstall` and go back to the AUR package at any time. Of course, there is *no* pressure to do this, and it is perfectly reasonable to wait on the AUR package to get updated if you don't feel comfortable running directly from source for any reason (running `sudo make install` is not always very good practice :sweat_smile:)

We can leave this issue open until you are able to test again :-)

**Comment by Dupiter-AU** [comment]
This is so weird. I ran ping for github, google and archlinux.org, and all but github returned a successful ping. 

![Screenshot_20221201_111732](https://user-images.githubusercontent.com/82429716/204942786-526408d2-f3a6-4edc-8843-4202ebfb11d7.png)

Could possibly be an ISP thing, but I find it very weird that only pings are non-functional, where I can still connect to github just fine in my browser.

Perhaps a `--force` flag would work to some extent, I know it definitely has when installing a mass selection of winetricks DLL's, heh.

If you by any chance know, how long would the AUR package usually take to update?

**Comment by sonic2kk** [comment]
That is really interesting, and it is strange that only pings would fail. However another user had that issue as far as I am aware, where only pinging would fail. I'll keep in mind that pinging can fail in this way, though the reason for the check is to see if the *script* can connect, so continuing with installation in a failure may not be very useful, since if the script can't `ping` it probably can't do the other fetching. Maybe at least the logging could be changed.

I am very sorry to say that right now I'm not sure if there's a better way forward to fix *this* issue, but the fixes pushed today should hopefully fix the offline stuff.

> If you by any chance know, how long would the AUR package usually take to update?

I actually would've expected it to be pretty much immediate, since I figured it just pulled in the latest git as new commits were made. You could maybe try running `yay -S steamtinkerlaunch-git` and seeing if that updates it?

The AUR package itself won't actually get "updated", so the last updated time on the AUR page is not a good indicator as far as I know. The PKGBUILD to my untrained eye doesn't have any useful hints on how it checks for updates. Maybe the [`pkgver`](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=steamtinkerlaunch-git#n44) section is used internally somewhere to check if the version on GitHub is newer than the cloned version? I am not too sure...

**Comment by Dupiter-AU** [comment]
I guess all that's left is to wait and see if the offline fixes work for manual installation. Thanks for the help!

> I actually would've expected it to be pretty much immediate, since I figured it just pulled in the latest git as new commits were made. You could maybe try running yay -S steamtinkerlaunch-git and seeing if that updates it?

I'm not so sure, the [`steamtinkerlaunch-git` AUR package](https://aur.archlinux.org/packages/steamtinkerlaunch-git) says it hasn't been updated since 2022-10-20.

**Comment by Dupiter-AU** [comment]
My cousin just pinged Github on their linux device and couldn't get a ping either, so it's not just on my end.

They have more experience with PCs and internet protocols, and they say that GitHub could be ignoring ICMP echo packets, which would explain why ping fails but browser connection works

**Comment by sonic2kk** [comment]
> I'm not so sure, the [`steamtinkerlaunch-git` AUR package](https://aur.archlinux.org/packages/steamtinkerlaunch-git) says it hasn't been updated since 2022-10-20.

This is for the PKGBUILD update, not the package itself. On the AUR, git packages update fairly infrequently, since they just pull in the code changes from GitHub and build them on your machine, the steps to pull in the package rarely need to change. That update time was for a repo URL update, when the project was handed over to me so the URLs needed to be updated to sonic2kk/steamtinkerlaunch instead of frostworx/steamtinkerlaunch. Hope that makes sense :-)

> My cousin just pinged Github on their linux device and couldn't get a ping either, so it's not just on my end.

That's very odd, it works on my end and others' ends. This is a strange issue.

> they say that GitHub could be ignoring ICMP echo packets, which would explain why ping fails but browser connection works

I'm not too familiar here but that sounds reasonable to me. I'm not sure how to fix that though... And I wonder why this only affects some devices?

**Comment by sonic2kk** [comment]
Looked into it and I found a few StackOverflow comments mentioning the 100% packet loss problem when using ping *specifically* for GitHub too, but they had no resolution.

I think a solution would be to not `return` if we can't ping then, though the rest of the downloading might fail. Simply having the warning might be enough.

I'll think about it some more and if I don't think it will cause too many problems with the rest of the flow, I'll just remove the return :-)

Thank you for the investigation on your end too!

**Comment by sonic2kk** [comment]
SteamTinkerLaunch v11.12.20221201 removes the `return` statement when GitHub cannot be pinged. STL may still be unable to download HMM from GitHub if it can't even ping it, but at least experimentally the `return` has been removed, so STL won't just cancel any downloading if we can't ping GitHub.

I am not sure yet how this will go, but it shouldn't start any major fires :-)

**Comment by Dupiter-AU** [comment]
Update: after the latest patches, HMM is now downloading, installing, and starting successfully. Although STL cannot ping GitHub, it can still successfully download from GitHub.

I'll link the successful start log just in case you want to take a double-check for any underlying bugs. Do note this log is from after the installation, because I went back to delete the download `.exe` to check if it was downloading correctly from GitHub.

[steamtinkerlaunch.log](https://github.com/sonic2kk/steamtinkerlaunch/files/10138515/steamtinkerlaunch.log)

I reckon I can close this issue now, so I'll close it with this comment

EDIT: I have noticed that HMM doesn't seem to have installed the d3d11 mod, so I'll just get that manually

EDIT 2: Uh oh I don't know where the HMM wine prefix is
![Screenshot_20221202_161731](https://user-images.githubusercontent.com/82429716/205228077-705bcab7-b6c1-492e-88ef-337d9b2fdcc4.png)

**Comment by sonic2kk** [comment]
Awesome stuff, I'm very glad! 😃

>  I have noticed that HMM doesn't seem to have installed the d3d11 mod, so I'll just get that manually

SteamTinkerLaunch does not install the D3D11 mod, it only installs the Winetrick to ensure compatibility.

Several users have been confused by this, is there any way to make this clearer? The D3D11 mod requires the `d3dcompiler_47` winetrick to be installed into the Sonic Generations prefix in order to function. STL only manages installing the Winetrick, not the mod. I am not sure how to make it clearer so I'm all ears for suggestions, the logging states it only installed `d3dcompiler_47` for compatibility, not that it installs the mod.

> Uh oh I don't know where the HMM wine prefix is

Installing 7zip or WinRAR doesn't seem to fix the issue. I've tested before with installing via Winetricks and HMM still errors out. So don't worry about installing it.

To fix this issue, extract the mod and use the HMM option to install from a folder. I'm not aware of anything on the STL side that could fix it unfortunately, maybe I could open an issue and ask the HMM Devs to check for and use `p7zip` on Linux.

But if you're curious, the HMM prefix is at `~/.config/steamtinkerlaunch/hedgemodmanager/compatdata/pfx` :-)

<hr>

I haven't checked your log yet but I'm very glad HMM is installed and working. Sorry for the confusion with the D3D11 mod, once again I'm very open on how to make what STL does clearer. Especially since if other mods for HMM compatible games require Winetricks, I'm planning to have STL install those too. I don't want users to think STL is somehow installing mods automatically.

Thank you so much for your patience! Have a great time modding :-)
